Add support for non-mainnet request cache validation thresholds

fselmo opened the pull request:
### What was wrong?

closes #3506

### How was it fixed?

- If non-mainnet ethereum, allow the use of an integer value for the request cache validation threshold. Pre-configure "safe" default values for some common non-mainnet chain ids based on their varied finality mechanisms.
- This integer value represents the number of seconds from time.now() that the request cache deems as a safe enough time window to allow the request to be cached.
- Update the tests to reflect these changes.
- Added comprehensive testing around the caching logic, affected endpoints, and all relevant providers.

Unrelated:

- Add a note to the ``request_mocker`` to make sure that mocked results
  are of the correct expected types based on JSON-RPC spec (e.g. hex
  strings instead of ints for numbers, etc.)
- Removed request caching decorator from base providers since they don't implement `make_request`. This makes typing tighter around the code base as well.
- IPCProvider did not have request caching turned on. This PR turns that on.

### Todo:

- [x] Add or update documentation related to these changes. This documentation should detail the underlying mechanism of how (and why) to check certain endpoints against a validation threshold so that users know:

  1. When to reach for caching in the first place
  2. If they reach for caching, clarify internal calls can be made to get the necessary validation data
  3. Know how they can tweak the validation threshold to lower the number of internal calls made (or set it to ``None`` to cache any and all requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)
- [x] Clean up commit history

#### Cute Animal Picture

```
     __      __
    /  \____/  \
   (    o  o    )
    \    ∇     /
     \  \___/  /
      \_______/
        |   |
```

fselmo committed:
Add support for non-mainnet request cache validation thresholds:

- If non-mainnet ethereum, allow the use of an integer value for the
  request cache validation threshold. Pre-configure "safe" default
  values for some common non-mainnet chain ids based on their
  varied finality mechanisms.

- This integer value represents the number of seconds from time.now()
  that the request cache deems as a safe enough time window to allow
  the request to be cached.

- Update the tests to reflect these changes.

Bonus:

- Add a note to the ``request_mocker`` to make sure that mocked results
  are of the correct expected types based on JSON-RPC spec (e.g. hex
  strings instead of ints for numbers, etc.)

fselmo committed:
Prevent recursion internally when caching requests; add tests:

- In order to prevent recursion, we need to turn off caching while we
  make internal calls to get the block information we need. Turn it back
  on via a ``finally`` so we are sure to turn it back on.

- Add some comprehensive tests for the now complex request caching
  functionality due to multi-chain support and internal-based
  caching threshold configuration.

fselmo committed:
Update request caching documentation

kclowes commented on web3/_utils/caching/caching_utils.py:
```diff
@@ -142,16 +173,60 @@ def is_cacheable_request(
+                    "Request cache validation threshold is set to "
+                    f"{current_threshold.value} "
+                    f"for chain with chain_id `{chain_id}` but this value only works "
+                    "on chain_id `1`. Setting to default value for chain_id "
```
This won't work with other POS chains? testnets?

fselmo replied:
Good catch. It should work with others as well. I think this messaging should be improved. Will fix 👍🏼

fselmo replied:
I think we can just remove that validation entirely. I was trying to issue a warning related to that value not being available, for friendliness, but that responsibility may have to be on the user since we can't realistically keep track of all chains that have it available.

kclowes commented on web3/_utils/caching/request_caching_validation.py:
```diff
@@ -31,99 +32,244 @@ def _error_log(
         _error_log(provider, e)
         return False
+    finally:
+        provider.cache_allowed_requests = True
```
lazy question: instead of always setting this back to true, I wonder if it makes sense to store the value of cache_allowed_requests coming in? for example, something like:
on line 46:
`cache_allowed_requests = provider.cache_allowed_requests`
and then here:
`provider.cache_allowed_requests = cache_allowed_requests`

fselmo replied:
I think we can for safety reasons but we should technically only get here if we are caching in the first place. Better to be safe though?

kclowes commented on web3/_utils/caching/request_caching_validation.py:
```diff
@@ -31,99 +32,244 @@ def _error_log(
-    blocknum = to_int(hexstr=block_id)
-    return is_beyond_validation_threshold(provider, blocknum)
 
+    blocknum = int(block_id, 16)
```
another lazy question: Is there a difference between what `to_int` is doing and `int` here?

fselmo replied:
I think it's more of an overhead than necessary? If we know that we're getting back a hex string from the client I though it'd be best to just make use of the `int` directly. Thoughts?

kclowes replied:
Cool, that makes sense to me!

kclowes commented on web3/_utils/module_testing/utils.py:
```diff
@@ -35,6 +35,14 @@ class RequestMocker:
     Context manager to mock requests made by a web3 instance. This is meant to be used
     via a ``request_mocker`` fixture defined within the appropriate context.
 
+    ************************************************************************************
+    Important: When mocking results, it's important to keep in mind the types that
+        clients return. For example, what we commonly translate to integers are returned
+        as hex strings in the RPC response and should be mocked as such for more
+        accurate testing.
+    ************************************************************************************
```
👍

kclowes commented on tests/core/caching-utils/test_request_caching.py:
```diff
@@ -260,30 +265,31 @@ def test_blocknum_validation_against_validation_threshold_when_caching(
         w3.manager.request_blocking(RPCEndpoint(endpoint), [block_id, False])
         cached_items = len(w3.provider._request_cache.items())
-        assert cached_items == 1 if should_cache else cached_items == 0
+        assert cached_items > 0 if should_cache else cached_items == 0
```
curious about this change?

fselmo replied:
oops, that was me playing around with some different ideas I think. Thought I had changed them all back. This was a result of not turning off caching when making requests to get validation blocks, etc. But it's not a good approach since it can enter recursion. I'll change this back, good catch 👍🏼

pacrob commented on docs/internals.rst:
```diff
@@ -119,48 +119,100 @@ Provider Configurations
+caching responses that depend on block data. By default, this option is configured
+to internal values deemed "safe" for the chain id you are connected to. If you are
+connected to mainnet Ethereum, this value is set to the ``finalized`` block number.
+If you are connected to another chain, this value is set to a time internal in seconds,
```
did you mean "is set internally to a time in seconds" or something like that?

fselmo replied:
Womp... "time interval". Will fix

fselmo committed:
Changes from comments on PR #3508:

- Fix typo in docs.
- Fix test check, be more precise and check cache == 1.
- Remove validation for threshold with chain_id == 1 since
  more chains than mainnet Ethereum can make use of ``finalized``
  and ``safe`` thresholds.
- Store and reset the current value of ``cache_allowed_requests`` rather than assuming ``True`` / ``False``

kclowes commented on tests/core/caching-utils/test_request_caching.py:
```diff
@@ -211,11 +215,11 @@ def test_all_providers_do_not_cache_by_default_and_can_set_caching_properties(pr
 ):
     w3 = Web3(
-        HTTPProvider(
+        BaseProvider(
```
more lazy questions: Is it just the HTTPProvider that caches? or should we maybe have a test that parametrizes the other providers too?

fselmo replied:
Nope, all providers should cache. I think that would be a good idea. Should we track it separately though?

fselmo replied:
We have a [test for caching properties](https://github.com/ethereum/web3.py/blob/46913500aba08113aad07c50b858aca18cc70f46/tests/core/caching-utils/test_request_caching.py#L167-L199) but we don't run all providers through the test suite currently.

kclowes replied:
> Nope, all providers should cache. I think that would be a good idea. Should we track it separately though?

Sure, that's fine by me if you'd rather track separately.

fselmo replied:
I added it in here instead, as a different commit.

kclowes reviewed (approved):
Mostly left lazy questions and small things, but looks good otherwise!

fselmo committed:
Test all providers for request caching; add caching to IPCProvider

fselmo committed:
Remove caching decorator from base providers; fix typing:

- Base providers don't implement ``make_request`` and so can't cache
  requests. Remove the request caching decorator from the base provider
  classes.

- Fix typing in the request caching utils, which should now be tighter,
  expecting an ``RPCEndpoint`` instead of a generic ``str``.

pacrob commented on web3/_utils/module_testing/utils.py:
```diff
@@ -99,7 +107,7 @@ def __init__(
     def __enter__(self) -> "Self":
         # mypy error: Cannot assign to a method
-        self.w3.provider.make_request = self._mock_request_handler  # type: ignore[method-assign]  # noqa: E501
+        self.w3.provider.make_request = self._mock_request_handler  # type: ignore
```
Why remove the type of ignore here?

fselmo replied:
Good point I think `__enter__` can keep `[method-assignment]` but I think `__exit__` had the issue. I think we can be more specific there and use `__assignment__`. This happened when I removed the request_caching decorators from the base provider classes. I think the typing ends up slightly off with `Arg()` wrappers around the args even though the type is essentially the same.

@pacrob are you good with narrowing the `__exit__` ignores to `[assignment]` and keeping `[method-assign]` here?

pacrob replied:
yup, sgtm

pacrob reviewed (approved):
just one q, otherwise lgtm!
